src/core/error.rs

```diff
@@ -0,0 +1,207 @@
+        }
+    }
+}
+
+impl<E, H, HI> Clone for ErrorDecorator<E, H, HI>
+where
+    E: Clone,
+    H: Clone,
+{
+    fn clone(&self) -> Self {
+        Self {
+            error_handler: self.error_handler.clone(),
+            handler: self.handler.clone(),
+            handler_input: PhantomData,
+        }
+    }
+}
+
+impl<E, H, HI> Handler<HandlerInput> for ErrorDecorator<E, H, HI>
+where
+    E: ErrorHandler + Clone + 'static,
+    H: Handler<HI> + 'static,
+    HI: TryFromInput,
+    HI::Error: 'static,
+    H::Output: Into<HandlerResult>,
+{
+    type Output = HandlerResult;
+    type Future = BoxFuture<'static, Self::Output>;
+
+    fn handle(&self, input: HandlerInput) -> Self::Future {
+        let handler = self.handler.clone();
+        let error_handler = self.error_handler.clone();
+        Box::pin(async move {
+            let future = HI::try_from_input(input);
+            match future.await {
+                Ok(Some(input)) => {
+                    let future = handler.handle(input);
+                    match future.await.into() {
+                        HandlerResult::Error(err) => {
+                            let future = error_handler.handle(err);
+                            HandlerResult::Error(future.await)
+                        }
+                        result => result,
+                    }
+                }
+                Ok(None) => HandlerResult::Continue,
+                Err(err) => {
+                    let future = error_handler.handle(Box::new(err));
+                    HandlerResult::Error(future.await)
+                }
+            }
+        })
+    }
+}
+
+/// Allows to process errors returned by handlers
+pub trait ErrorHandler: Send {
+    /// A future returned by `handle` method
+    type Future: Future<Output = HandlerError> + Send;
+
+    /// Handles a errors
+    ///
+    /// # Arguments
+    ///
+    /// * err - An error to handle
+    ///
+    /// You need to return the error in order to pass it to subsequent handler
+    fn handle(&self, err: HandlerError) -> Self::Future;
+}
+
+impl<H, F> ErrorHandler for H
+where
+    H: Fn(HandlerError) -> F + Send,
+    F: Future<Output = HandlerError> + Send,
+{
+    type Future = F;
+
+    fn handle(&self, err: HandlerError) -> Self::Future {
+        (self)(err)
+    }
+}
+
+/// Writes an error to log
+pub struct LoggingErrorHandler;
+
+impl ErrorHandler for LoggingErrorHandler {
+    type Future = BoxFuture<'static, HandlerError>;
+
+    fn handle(&self, err: HandlerError) -> Self::Future {
+        Box::pin(async move {
+            log::error!("An error has occurred: {}", err);
+            err
+        })
+    }
+}
+
+/// Error decorator shortcuts
+pub trait ErrorExt<E, HI>: Sized {
+    /// Shortcut to create a new error decorator (`handler.error(error_handler)`)
+    ///
+    /// # Arguments
+    ///
+    /// * error_handler - An error handler
+    fn error(self, error_handler: E) -> ErrorDecorator<E, Self, HI> {
```

Comment:
Is it better to name it like `on_error`? It's not like we always want error, it acts like predicate